File: src/main/java/com/bdev/hogwarts_api/rest_service/teacher/TeacherRestServiceImpl.java
package com.bdev.hogwarts_api.rest_service.teacher;

import acropollis.municipali.security.common.dto.MunicipaliUserInfo;
import com.bdev.hogwarts_api.data.dto.teacher.Teacher;
import com.bdev.hogwarts_api.exceptions.http.HttpEntityIllegalStateException;
import com.bdev.hogwarts_api.exceptions.http.HttpEntityNotFoundException;
import com.bdev.hogwarts_api.service.teacher.TeacherService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class TeacherRestServiceImpl implements TeacherRestService {
    @Autowired
    private TeacherService teacherService;

    @Transactional(readOnly = true)
    @Override
    public List<Teacher> getAllTeachers(
            MunicipaliUserInfo userInfo
    ) {
        return teacherService.getAllTeachers();
    }

    @Transactional(readOnly = true)
    @Override
    public Teacher getTeacherById(
            MunicipaliUserInfo userInfo,
            long id
    ) {
        return teacherService.getTeacherById(id)
                .orElseThrow(() -> new HttpEntityNotFoundException("Teacher with id '%d' does not exist", id));
    }

    @Transactional(readOnly = true)
    @Override
    public Teacher getTeacherByLogin(
            MunicipaliUserInfo userInfo,
            String login
    ) {
        return teacherService.getTeacherByLogin(login)
                .orElseThrow(() -> new HttpEntityNotFoundException("Teacher with login '%s' does not exist", login));
    }

    @Transactional
    @Override
    public long createTeacher(
            MunicipaliUserInfo userInfo,
            Teacher teacher
    ) {
        if (teacher.getId() != null) {
            throw new HttpEntityIllegalStateException("Teacher id should be null during creation");
        }

        return teacherService.createTeacher(teacher);
    }

    @Transactional
    @Override
    public void editTeacher(
            MunicipaliUserInfo userInfo,
            Teacher teacher
    ) {
        if (teacher.getId() == null) {
            throw new HttpEntityIllegalStateException("Teacher id should be non-null during edit");
        }

        if (!teacherService.exists(teacher.getId())) {
            throw new HttpEntityNotFoundException("Teacher with id '%d' does not exist", teacher.getId());
        }

        teacherService.editTeacher(teacher);
    }

    @Transactional
    @Override
    public void deleteTeacher(
            MunicipaliUserInfo userInfo,
            long id
    ) {
        if (!teacherService.exists(id)) {
            throw new HttpEntityNotFoundException("Teacher with id '%d' does not exist", id);
        }

        teacherService.deleteTeacher(id);
    }
}

File: src/main/java/com/bdev/hogwarts_api/rest_service/teacher/TeacherRestServiceImpl.kt
package com.bdev.hogwarts_api.rest_service.teacher

import acropollis.municipali.security.common.dto.MunicipaliUserInfo
import com.bdev.hogwarts_api.data.dto.teacher.Teacher
import com.bdev.hogwarts_api.exceptions.http.HttpEntityIllegalStateException
import com.bdev.hogwarts_api.exceptions.http.HttpEntityNotFoundException
import com.bdev.hogwarts_api.service.teacher.TeacherService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
open class TeacherRestServiceImpl : TeacherRestService {
    @Autowired
    private lateinit var teacherService: TeacherService

    @Transactional(readOnly = true)
    override fun getAllTeachers(userInfo: MunicipaliUserInfo): List<Teacher> {
        return teacherService.getAllTeachers()
    }

    @Transactional(readOnly = true)
    override fun getTeacherById(userInfo: MunicipaliUserInfo, id: Long): Teacher {
        return teacherService
                .getTeacherById(id)
                ?: throw HttpEntityNotFoundException("Teacher with id '%d' does not exist", id)
    }

    @Transactional(readOnly = true)
    override fun getTeacherByLogin(userInfo: MunicipaliUserInfo, login: String): Teacher {
        return teacherService
                .getTeacherByLogin(login)
                ?: throw HttpEntityNotFoundException("Teacher with login '%s' does not exist", login)
    }

    @Transactional
    override fun createTeacher(userInfo: MunicipaliUserInfo, teacher: Teacher): Long {
        if (teacher.id != null) {
            throw HttpEntityIllegalStateException("Teacher id should be null during creation")
        }

        return teacherService.createTeacher(teacher)
    }

    @Transactional
    override fun editTeacher(
            userInfo: MunicipaliUserInfo,
            teacher: Teacher
    ) {
        val teacherId = teacher.id
                ?: throw HttpEntityIllegalStateException("Teacher id should be non-null during edit")

        if (!teacherService.exists(teacherId)) {
            throw HttpEntityNotFoundException("Teacher with id '%d' does not exist", teacherId)
        }

        teacherService.editTeacher(teacher)
    }

    @Transactional
    override fun deleteTeacher(userInfo: MunicipaliUserInfo, id: Long) {
        if (!teacherService.exists(id)) {
            throw HttpEntityNotFoundException("Teacher with id '%d' does not exist", id)
        }

        teacherService.deleteTeacher(id)
    }
}
